First draft on a review screen tutorial / guide

BryceStevenWilley opened the pull request:
Very much in first draft mode. Going to walk through it myself next week when trying to make a review screen for an interview, and will make any adjustments necessary. Then I'll un-draft this PR. Feel free to review if you want to before hand though.

nonprofittechy commented on docs/writing_review_screen.md:
````diff
@@ -0,0 +1,240 @@
+      - birthdate
+    ```
+
+### Step 3: Edits to the Review Screen Generator
+
+You'll want to change some aspects of the generated review screen. Note that if any errors happen in a specific edit element on the review screen, that element won't appear. It also won't appear if the attribute or variable to be edited doesn't exist. You should make sure that it does.
+
````
I'd add sub-headings to this section. The call to action of "Edit the output" is clear enough, but it doesn't say what we'll actually do functionally in this section. Can that be added to the top level heading, too?

nonprofittechy commented on docs/writing_review_screen.md:
````diff
@@ -0,0 +1,240 @@
+```
+
+2. Remove tables for attachments and other non-people objects, if any were generated. Tables are only needed for information that the user enters themselves, which the review screen generator can't determine yet.
+3. Remove edit questions about kickout screens. Sometimes, you will have questions that do set a variable in the interview, but it's only use is to kickout a user if their problem isn't covered by the form. Here's an example of one:
+
````
I'd make the heading "Remove questions you don't want the user to edit" and make the kickout screen an example.

nonprofittechy commented on docs/writing_review_screen.md:
```diff
@@ -0,0 +1,240 @@
+If editing the value of a variable like this would only lead the person to a kickout screen and not change any other logic of the interview, then the generated review section should be removed.
+
+
+### Step 4: Things to add to the review screen
+
```
This isn't such a helpful heading. What about directly describing the things to add or customize?

E.g., "Add logic for follow-up questions when an answer is changed by the user"

nonprofittechy commented on docs/writing_review_screen.md:
````diff
@@ -0,0 +1,240 @@
+```
+
+2. As mentioned in step 2, if a variable doesn't exist or any errors occur when showing a section of the review screen, that whole section won't be shown. If some of the fields on a screen or optional, then their absence can cause their section of the review screen to not appear. Use `showifdef` on optional fields, or fields that use `show if`, to prevent this issue.
+3. While writing the review screen, you will likely have to redo your interview logic, so that it doesn't make assumptions about the presence or absence of variables, just their values, which I'd encourage. From the docassemble documentation:
+
````
This is a great point. I think it needs its own heading. Maybe:

Defensive interview writing: Writing your interview so that answers can be changed by the user

nonprofittechy reviewed (approved):
Nice start! I have some ideas that hopefully can help, mostly about structure, headings, and formatting to make it easier to jump to content an author needs.

nonprofittechy reviewed (changes requested):
See previous review

BryceStevenWilley commented on docs/writing_review_screen.md:
```diff
@@ -0,0 +1,240 @@
+    Each element in the review can have the following dictionary attributes:
+
+    * `Edit` adds an "Edit" button next to the information on the review screen. It's value is the name of the variable that will be edited when the button is clicked.
+    * `label` lets you change the text on the "Edit" button
+    * `fields` lets you add multiple fields that are re-asked to the user for a section on the review screen.
```
I had this in because I thought that `Edit` only took single variables, not the lists that allowed for `recompute` in them, which is something that the generated review screen doesn't use, and is very necessary for correctness in review screens. Pleasantly surprised that `Edit` takes lists of things, so I'll make this simpler

BryceStevenWilley commented on docs/writing_review_screen.md:
````diff
@@ -0,0 +1,240 @@
+      * `recompute` lets you also re-calculate other variables, not necessarily re-asking them, but letting them be filled by `code` blocks. These won't be generated, but will come in handy later.
+    * `button` is the actual content of that section of the review screen, similar to a `subquestion` attribute.
+
+3. The revisit screens. These blocks look like this:
+
+    ```yml
+    id: revisit users
+    continue button field: users.revisit
+    question: |
+      Edit your answers about Users
+    subquestion: |
+      ${ users.table }
+
+      ${ users.add_action() }
+    ---
+    ...
+    ```
+
+    Each block has the table and lets people edit items in the lists.
+
+4. The generated tables for the lists.
+
+    ```yml
+    table: children.table
+    rows: children
+    columns:
+      - Birthdate: |
+          row_item.birthdate if hasattr(row_item, 'birthdate') else ''
+    edit:
+      - birthdate
+    ```
+
````
> I think this should start by explaining something about the purpose--that these two sections are about how you can edit lists using the review screen approach, and why it might be useful to do so

I agree that more context is needed, but since these are already a part of the "What the generator makes for you" section, I'll stick with describing what they'll see in their generated review screen before going into the what or why.

> which are just lists of people, and maybe the ones that they can delete. Probably we should delete those in the review screen generator though.

Reminds me that there's some upstream changes I need to make to Assembly Line, we don't have any `revisit` screens for the people lists that we define, which we should definitely have.

BryceStevenWilley replied:
Made those upstream changes (PR incoming), and I took a shot at adding more context to this section.

BryceStevenWilley commented:
Thanks for the review! I addressed some of the comments. The latter portion of the guide was more free form because I'm sure I'll have to add more types of changes to make to the review screen there. I'm gonna do the test run on a MLH interview this week, add whatever I need to to full things in, and try to integrate the rest of the comments then as well.

Note to myself but I do want to link to the docassemble docs a lot more, but I'll have to add anchor links in the docs so I can link to specific parts of the page, and not just the 1.5k word section on "Customizing the display of review options".

BryceStevenWilley commented on docs/writing_review_screen.md:
```diff
@@ -0,0 +1,240 @@
+### Step 2: Understanding the Generated Review Screen
+
+A first quick note is that the generated YAML might look different than how you're used to writing YAML files. The generated YAML uses lots of `|-`,
+which keeps the full text after, but removes the last newline. In short, something like this:
```
I open with it, because the examples that I use in here are the concise versions, and this is explain to readers that these things are the same. IMO this core YAML syntax is stuff that is assumed in the docassemble docs and not covered well elsewhere, so it's important to have little tidbits like this.

I shortened it up a lot in this edit and I think the shorter version in an admonition is a good compromise.

BryceStevenWilley commented on docs/writing_review_screen.md:
```diff
@@ -0,0 +1,240 @@
+Large changes to an interview's logic will require large changes to the review screen as well. Therefore, it's best to have mostly completed
+your interview before starting to make the review screen.
+
+### Step 1: ALDashboard Review Screen Generator
```
> Some bullets or a numbered list describing the steps at a high level might help people jump to the part that has the information that they need.

Is the "Overview" off to the right what you are thinking of here? Still going to improve these section header a bit which should help too.

![Screenshot from 2023-12-15 10-31-16](https://github.com/SuffolkLITLab/docassemble-AssemblyLine-documentation/assets/6252212/36581cd5-5bad-4a58-8ae6-8d6d2b123ef2)

nonprofittechy replied:
I think broken down into more detail, and focused on the substance of each step might be useful than the current headings you screenshotted. But even so a quick bulleted list inline will help.

Something like:

This guide will cover:

* What a review screen is for
* The different pieces and capabilities of a review screen 
* How to use Assembly Line tools to speed up authoring a review screen
* How to read, understand, and modify the output of the Assembly Line review screen generator
* Common review screen tasks that are not handled by the generator tool
* How to write your interview in a way that allows the user's edits to be reflected in the final document

This is written as a tutorial you can follow, but you can also jump right to the topic that you need help with.

BryceStevenWilley commented on docs/writing_review_screen.md:
```diff
@@ -0,0 +1,240 @@
+Large changes to an interview's logic will require large changes to the review screen as well. Therefore, it's best to have mostly completed
+your interview before starting to make the review screen.
+
+### Step 1: ALDashboard Review Screen Generator
```
> I'd suggest explaining what the user's task is when making a review screen

Do you have anything in mind specifically here? I tried to cover the UX perspective above (users need a way to edit answers, and DA's "undo as back" doesn't cut it). I don't want this doc to scope creep to becoming a concepts explainer, technical guide, and a tutorial, I want it to be a tutorial first and foremost. And most of what the review generator doesn't do is in the technical details that are hard to understand until you've used / seen a review screen

nonprofittechy replied:
Understood. Tutorials are definitely helpful, and people need a place to get started. The review screen section of the main documentation is so hard to use though, that a general explainer and resource could be helpful. But understood if you don't want to make this that thing.

I was not thinking of a full API reference. Maybe something like this:

Review screens let your user change their mind about their previous answers. Like the list of fields in a `question` screen, one review screen can let the user review many answers. This list of answers goes under the heading `review` instead of `fields`. Each answer you want to let your user change includes:

* A button, with a label of your choice (we usually use "Edit")
* A field name or a list of field names that you want the user to revisit when they click the edit button
* Some preview of the user's current answer. It's common to repeat or paraphrase the question from the interview and show the answer next to the question.

You can do more advanced things with review screens, too:
* If you would like the user to be able to jump to another screen (perhaps to edit a list of items), you can make the field name point to another question screen, labeled with an  `event`. This strategy is usually used to edit lists of people with a `revisit` block.
* You can trigger follow-up code to run after the user makes a change. For example: you can trigger a code block to get recomputed, delete the value of another answer, or trigger another follow-up screen.

BryceStevenWilley commented:
Shoot, sorry, I had made some more progress on this but wasn't able to finish it entirely. Aside from general polish, there was still a section I wanted to add about using code blocks to re-evaluate whole chunks of logic, which is unfortunately necessary with really complicated dependent logic (Like [what we do in the ALAffidavit](https://github.com/SuffolkLITLab/docassemble-ALAffidavitOfIndigency/blob/5c908e47a06ee941c2bbe8395e5292eb3d2dfcf5/docassemble/ALAffidavitOfIndigency/data/questions/review_page.yml#L187-L193), and [a more complicated version of that in MLH code](https://github.com/mplp/docassemble-Simpmotion/blob/76371898436c2438a297873f8ecf3dc33389d008/docassemble/Simpmotion/data/questions/simpmotion.yml#L1218-L1243) when re-emptying and re-filling user lists), and also really error prone.

Feel free to merge as is though, just wanted to document the current state it's in. Might need a few edits and stuff, so go wild if need be.

nonprofittechy reviewed (approved):
LG for now! We can keep adding new content over time. Thanks for your work on this!
